### PR TITLE
[Test] Make sigmoid_add extension test device-agnostic in test_cpp_extensions_aot.py

### PR DESCRIPTION
**Pull request by vishalgoyal316**
## Summary

Merge `test_cuda_extension` and `test_sycl_extension` into a single `test_sigmoid_add_extension` parametrized over the backends listed in `_SIGMOID_ADD_BACKENDS` (cpu, mps, cuda, xpu). Each backend appears as its own test entry in CI so pass/skip/fail is visible per device instead of being collapsed into one `subTest` result.

## Changes

- Add module-level `_SIGMOID_ADD_BACKENDS` dict mapping each device type to its pre-built extension module:
  - cpu and mps share the ATen-dispatched `.cpp` extension
  - cuda and xpu use their dedicated kernel extensions from `setup.py`
- Replace the `subTest` loop with `@parametrize("device_type", ...)`; look up `module_name` from the dict inside the test
- Use `torch.get_device_module(device_type).is_available()` for runtime availability checks
- Skip (not fail) when a device is unavailable at runtime, or when XPU is tested under `USE_NINJA=0` (the sycl extension is not built in the `test_cpp_extensions_aot_no_ninja` target)
- Add `instantiate_parametrized_tests(TestCppExtensionAOT)` so parametrized cases are expanded at module level
- Remove unused `TEST_XPU` import

## Also in this PR

- Split the AOT test class by device scope (`TestCppExtensionAOTDevice` / `TestCppExtensionAOTCUDA` / `TestCppExtensionAOTMPS`) and annotate every class in the file with `hw_classification`

## Not changed behaviorally

- `test_mps_extension` (Metal shader + MPS stream/encoder APIs specific to the `.mm` extension), `test_cublas_extension` / `test_cusolver_extension` (NVIDIA library link tests), `test_cuda_dlink_libs` (CUDA device-linking build-system test), and the `TestMAIATensor` / `TestRNGExtension` / `TestPybindTypeCasters` / `TestTorchLibrary` classes keep their existing test logic; the final commit only regrouped them into the device-scoped classes and added `hw_classification` annotations

## Test plan

```bash
python run_test.py -i test_cpp_extensions_aot_ninja -k test_sigmoid_add_extension
python run_test.py -i test_cpp_extensions_aot_no_ninja -k test_sigmoid_add_extension
```
---

Authored with an AI assistant.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @mcarilli @ptrblck @leslie-fang-intel @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @xmfan @aditvenk

### CODE REVIEW

**Comment by vishalgoyal316** [comment]
cc: @mansiag05 @RiyaP2508 @fffrog

File: test/test_cpp_extensions_aot.py [hunk 1]
```diff
@@ -85,17 +95,29 @@ def test_backward(self):
...
-        x = torch.zeros(100, device="cuda", dtype=torch.float32)
-        y = torch.zeros(100, device="cuda", dtype=torch.float32)
-
-        z = cuda_extension.sigmoid_add(x, y).cpu()
-
-        # 2 * sigmoid(0) = 2 * 0.5 = 1
-        self.assertEqual(z, torch.ones_like(z))
+    def test_sigmoid_add_extension(self):
+        for device_type, module_name in _SIGMOID_ADD_BACKENDS:
+            with self.subTest(device_type=device_type):
```
Comment (jbschlosser) [inline test/test_cpp_extensions_aot.py]: stylistic nit: instead of `subTest`, I suggest using `@parametrize` over `device_type` to be more idiomatic

**Review by jbschlosser — COMMENTED** [review]
One style nit

**Comment by jbschlosser** [comment]
@claude review this PR and summarize the changes

File: test/test_cpp_extensions_aot.py [hunk 2]
```diff
@@ -85,17 +95,29 @@ def test_backward(self):
...
+                    if is_available is not None and not is_available():
+                        raise unittest.SkipTest(f"{device_type} not available")
+                    if device_type == "xpu" and os.getenv("USE_NINJA", "0") == "0":
+                        raise unittest.SkipTest(
+                            "sycl extension requires ninja to build"
+                        )
+                try:
+                    ext = importlib.import_module(module_name)
+                except ImportError:
+                    raise unittest.SkipTest(f"{module_name} not built") from None
```
Comment (jbschlosser) [inline test/test_cpp_extensions_aot.py]: I think we should hard-error here instead, as not being able to import means something is wrong the test is trying to check!

File: test/test_cpp_extensions_aot.py [hunk 3]
```diff
@@ -85,17 +95,29 @@ def test_backward(self):
...
+                    device_mod = getattr(torch, device_type, None)
+                    if device_mod is None:
+                        raise unittest.SkipTest(f"{device_type} is not registered")
+                    is_available = getattr(device_mod, "is_available", None)
+                    if is_available is not None and not is_available():
+                        raise unittest.SkipTest(f"{device_type} not available")
+                    if device_type == "xpu" and os.getenv("USE_NINJA", "0") == "0":
+                        raise unittest.SkipTest(
+                            "sycl extension requires ninja to build"
+                        )
```
Comment (jbschlosser) [inline test/test_cpp_extensions_aot.py]: I'd hard error here instead. The test shouldn't silently be skipped because of a missing build tool

File: test/test_cpp_extensions_aot.py [hunk 4]
```diff
@@ -85,17 +95,29 @@ def test_backward(self):
...
-
-        # 2 * sigmoid(0) = 2 * 0.5 = 1
-        self.assertEqual(z, torch.ones_like(z))
+    def test_sigmoid_add_extension(self):
+        for device_type, module_name in _SIGMOID_ADD_BACKENDS:
+            with self.subTest(device_type=device_type):
+                if device_type != "cpu":
+                    device_mod = getattr(torch, device_type, None)
+                    if device_mod is None:
+                        raise unittest.SkipTest(f"{device_type} is not registered")
```
Comment (jbschlosser) [inline test/test_cpp_extensions_aot.py]: In general, I don't think we should skip at all, but rather hard error if anything goes wrong

**Comment by vishalgoyal316** [comment]
Sorry, while rebasing, wrong commits are pushed. fixing it.

**Comment by vishalgoyal316** [comment]
Hello @jbschlosser, thanks for the review feedback.
I have addressed the comments, and request you to review it.
(while rebasing of branch, by mistake many commits added to this PR, and later I fixed it, could you please help to clean reviewer and label section if required. Sorry for the inconvenience).

File: test/test_cpp_extensions_aot.py [hunk 5]
```diff
@@ -85,15 +97,22 @@ def test_backward(self):
...
-        x = torch.zeros(100, device="cuda", dtype=torch.float32)
-        y = torch.zeros(100, device="cuda", dtype=torch.float32)
-
-        z = cuda_extension.sigmoid_add(x, y).cpu()
-
+    @parametrize("device_type,module_name", list(_SIGMOID_ADD_BACKENDS))
+    def test_sigmoid_add_extension(self, device_type, module_name):
+        if device_type != "cpu":
+            device_mod = getattr(torch, device_type, None)
+            if device_mod is None:
```
Comment (aorenste) [inline test/test_cpp_extensions_aot.py]: [cleanup] This block hand-rolls `torch.get_device_module()`, which does the identical `getattr(torch, device_type, None)` + None-check internally (`torch/__init__.py`) but with an accurate message ("does not have a corresponding module registered as 'torch.X'"). The current "not compiled into this PyTorch build" wording describes a condition this check can't reach: a not-compiled-in backend still ships its `torch.<device>` submodule (e.g. `torch.mps` exists on a Linux build where `torch.backends.mps.is_built()` is False), so it takes the `is_available()` skip; `device_mod is None` only happens for an unknown/typo device name.

Since `get_device_module("cpu")` returns `torch.cpu` and `torch.cpu.is_available()` is always `True`, the `device_type != "cpu"` guard is also unnecessary -- cpu passes the availability check and isn't xpu, so it falls straight through. The whole block collapses to:

```python
        if not torch.get_device_module(device_type).is_available():
            raise unittest.SkipTest(f"{device_type} not available")
        if device_type == "xpu" and os.getenv("USE_NINJA", "0") == "0":
            raise unittest.SkipTest("sycl extension requires ninja to build")
```

**Review by aorenste — APPROVED** [review]
(Reviewed by me, assisted by AI)

**Review by jeffdaily — CHANGES_REQUESTED** [review]
Three items, one of which is just aorenste's cleanup that never got applied.

`@parametrize("device_type,module_name", ...)` parametrizes over both, so the module path lands in the generated test IDs: `test_sigmoid_add_extension_device_type_cuda_module_name_torch_test_cpp_extension_cuda`. `module_name` is a pure function of `device_type`, and jbschlosser's suggestion was to parametrize over `device_type` alone. Make `_SIGMOID_ADD_BACKENDS` a dict and look the module up in the body. These IDs are what the disable-test bot, `-k` selection and flaky tracking key on, so renaming later resets that history.

aorenste's 2026-07-21 comment on the availability block still applies verbatim to the current head. His collapsed version is right; take it as-is.

On jbschlosser's skip-vs-hard-error comments, which that collapsed version leaves as skips: keep the skips, at least for ninja. `test/cpp_extensions/setup.py` only builds `.sycl` when `USE_NINJA=1`, so hard-erroring would fail every job where the extension was never built rather than surfacing a real problem. Worth stating on the thread so the point is closed rather than left open.

The description still describes the `subTest` implementation and claims the check works for any registered backend including privateuse1; only the four table entries run. It becomes the commit message.

Review drafted with AI assistance.

**Comment by vishalgoyal316** [comment]
> Three items, one of which is just aorenste's cleanup that never got applied.
> 
> `@parametrize("device_type,module_name", ...)` parametrizes over both, so the module path lands in the generated test IDs: `test_sigmoid_add_extension_device_type_cuda_module_name_torch_test_cpp_extension_cuda`. `module_name` is a pure function of `device_type`, and jbschlosser's suggestion was to parametrize over `device_type` alone. Make `_SIGMOID_ADD_BACKENDS` a dict and look the module up in the body. These IDs are what the disable-test bot, `-k` selection and flaky tracking key on, so renaming later resets that history.
> 
> aorenste's 2026-07-21 comment on the availability block still applies verbatim to the current head. His collapsed version is right; take it as-is.
> 
> On jbschlosser's skip-vs-hard-error comments, which that collapsed version leaves as skips: keep the skips, at least for ninja. `test/cpp_extensions/setup.py` only builds `.sycl` when `USE_NINJA=1`, so hard-erroring would fail every job where the extension was never built rather than surfacing a real problem. Worth stating on the thread so the point is closed rather than left open.
> 
> The description still describes the `subTest` implementation and claims the check works for any registered backend including privateuse1; only the four table entries run. It becomes the commit message.
> 
> Review drafted with AI assistance.

Hello @jeffdaily, @aorenste, Thanks for the feedback. I kept the SkipTest for the XPU + USE_NINJA=0 case rather than hard-failing.

run_test.py runs this file under two separate CI targets — test_cpp_extensions_aot_ninja (USE_NINJA=1) and test_cpp_extensions_aot_no_ninja (USE_NINJA=0). The sycl/XPU extension is only built when ninja is enabled, so on the no-ninja target the XPU parametrized case is expected to skip, not fail. That matches the existing pattern in test_cuda_dlink_libs, which also skips when USE_NINJA=0.

A hard error there would make test_cpp_extensions_aot_no_ninja red whenever XPU is compiled in, even though that configuration is intentional. The ninja-enabled target is where we actually exercise the XPU path.

Rest of the comments also addressed, please help to review again, thanks again!

File: test/test_cpp_extensions_aot.py [hunk 6]
```diff
@@ -85,15 +97,17 @@ def test_backward(self):
...
-    @unittest.skipIf(not TEST_CUDA, "CUDA not found")
-    def test_cuda_extension(self):
-        import torch_test_cpp_extension.cuda as cuda_extension
-
-        x = torch.zeros(100, device="cuda", dtype=torch.float32)
-        y = torch.zeros(100, device="cuda", dtype=torch.float32)
-
-        z = cuda_extension.sigmoid_add(x, y).cpu()
-
+    @parametrize("device_type", list(_SIGMOID_ADD_BACKENDS))
```
Comment (aorenste) [inline test/test_cpp_extensions_aot.py]: [cleanup] My agent says:
> ```python
> @parametrize("device_type", list(_SIGMOID_ADD_BACKENDS), name_fn=lambda d: d)
> ```
>
> CI is generating `test_sigmoid_add_extension_device_type_cpu` / `_cuda` / `_mps` / `_xpu` (job `98199307192`). Extending jeffdaily's point that these IDs are what the disable-test bot, `-k` selection and flaky tracking key on: the `device_type_` infix is still redundant, and renaming later resets that history.

**Review by mansiag05 — APPROVED** [review]

**Review by jeffdaily — APPROVED** [review]
Approving. All items from my last review are addressed at 75f7c0a99: device_type-only parametrization with the dict lookup keeps the test IDs stable, the availability block is the collapsed form, the ninja-skip rationale is closed on the thread, and a final sweep found no remaining open request from any reviewer. I refreshed the description to cover the final class-split/hw_classification commit so the squash message matches the diff.

The failing checks are all verified unrelated: the xpu shards fail on the same inductor/test_compile_subprocess pickle error that has been red on trunk since 09-05, the symm-mem jobs run only symmetric-memory tests, pallas-tpu is unstable-labeled, and torchtitan is the long-standing trunk failure. None of the failed jobs execute test_cpp_extensions_aot.

Analysis assisted by AI.

**Comment by jeffdaily** [comment]
@pytorchbot merge -i

All failing checks are verified unrelated: xpu shards match the trunk-wide test_compile_subprocess pickle failure red on main since 09-05, symm-mem jobs do not run this file, pallas-tpu is unstable, torchtitan is the known trunk failure.

Analysis assisted by AI.
